### PR TITLE
fixes #3099 - Adding parameters to locations and organizations.

### PR DESCRIPTION
**Pull request by ohadlevy**
adds both API and UI.

two known issues:
- Audit message sometimes is not clear, not sure yet how to fix.
- once adding parameters to the many tabs under orgs/locations, it drops a
  line

### CODE REVIEW

**Comment by ohadlevy** [comment]
looks like a random jenkins failure, [test] again

File: app/controllers/api/v2/parameters_controller.rb [hunk 1]
```diff
@@ -48,10 +56,14 @@ def show
...
       api :POST, "/domains/:domain_id/parameters/", "Create a nested parameter for domain"
       api :POST, "/operatingsystems/:operatingsystem_id/parameters/", "Create a nested parameter for operating system"
+      api :POST, "/locations/:location_id/parameters/", "Create a nested parameter for location"
+      api :POST, "/organizations/:organization_id/parameters/", "Create a nested parameter for organization"
       param :host_id, String, :desc => "id of host"
       param :hostgroup_id, String, :desc => "id of hostgroup"
       param :domain_id, String, :desc => "id of domain"
       param :operatingsystem_id, String, :desc => "id of operating system"
+      param :location_id, String, :desc => "id of location"
+      param :organization_ida, String, :desc => "id of organization"
```
Comment (domcleal) [inline app/controllers/api/v2/parameters_controller.rb]: id, not ida

File: app/controllers/api/v2/parameters_controller.rb [hunk 2]
```diff
@@ -98,7 +118,8 @@ def destroy
       api :DELETE, "/hostgroups/:hostgroup_id/parameters", "Delete all nested parameters for hostgroup"
       api :DELETE, "/domains/:domain_id/parameters", "Delete all nested parameters for domain"
       api :DELETE, "/operatingsystems/:operatingsystem_id/parameters", "Delete all nested parameters for operating system"
-
+      api :DELETE, "/locations/:location_id/parameters", "Delete all nested parameter for location"
+      api :DELETE, "/organizations/:organization_id/parameters", "Delete all nested parameter for organization"
```
Comment (domcleal) [inline app/controllers/api/v2/parameters_controller.rb:122]: don't organization_id etc need to be defined here?

File: app/models/host/managed.rb [hunk 3]
```diff
@@ -338,6 +338,14 @@ def host_inherited_params include_source = false
     domain.domain_parameters.each {|p| hp.update Hash[p.name => include_source ? {:value => p.value, :source => :domain} : p.value] } unless domain.nil?
     # read OS parameters
     operatingsystem.os_parameters.each {|p| hp.update Hash[p.name => include_source ? {:value => p.value, :source => :os} : p.value] } unless operatingsystem.nil?
+    if SETTINGS[:locations_enabled]
```
Comment (domcleal) [inline app/models/host/managed.rb]: Jon on the ticket suggested moving this up, closer to global - what do you think?

File: app/models/parameters/organization_parameter.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  audited :except => [:priority], :associated_with => :organization, :allow_mass_assignment => true
+  validates :name, :uniqueness => {:scope => :reference_id}
+
+  private
+  def enforce_permissions operation
+    # We get called again with the operation being set to create
+    return true if operation == "edit" and new_record?
+    return true if User.current.allowed_to?("#{operation}_organizations".to_sym)
+
+    errors.add(:base, _("You do not have permission to %s this Organization parameter") % operation)
```
Comment (domcleal) [inline app/models/parameters/organization_parameter.rb]: lower case "O", ditto below for location

**Comment by domcleal** [comment]
There aren't many tests ;)

**Comment by domcleal** [comment]
[test]

File: app/controllers/api/v2/parameters_controller.rb (hunk 2 above)
Comment (ohadlevy) [inline app/controllers/api/v2/parameters_controller.rb:122]: I'm not sure, none of them were defined, @isratrade ?

**Comment by ohadlevy** [comment]
@domcleal updated with basic tests

**Comment by domcleal** [comment]
Thanks @ohadlevy, merged as 2de7832 for Foreman 1.4.0.

**Pull request closed without merging**